ui/lib/CancelStream.tsx

```diff
@@ -1,7 +1,116 @@
+  const [inProgress, setInProgress] = useState(false);
+  // Get the minimum stream life, before which a stream cannot be Canceled
+  const minStreamLife = Number(props.stream.streamManager.MIN_STREAM_LIFE);
+  // Get the starting time of a stream
+  const [startTime, setStartTime] = useState<number>(0);
+  // Get the current time to be able to calculate remaining time before cancellability
+  const [currentTime, setCurrentTime] = useState(Date.now() / 1000);
+
+  // Check if the stream is cancellable and set the button state accordingly.
+  useEffect(() => {
+    // Check if the stream is cancellable and set the button state accordingly
+    const checkStreamCancelable = async () => {
+      try {
+        const isCancelable = await props.stream.isCancelable();
+        setButtonEnabled(isCancelable);
+        if (isCancelable) {
+          clearInterval(interval);
+        }
+      } catch (error) {
+        console.error("Error checking stream cancellability:", error);
+      }
+    };
+
+    // Set interval to check cancellability every 10 seconds
+    const interval = setInterval(checkStreamCancelable, 10000);
+
+    // Clean up the interval when the component unmounts or isButtonEnabled is true
+    return () => clearInterval(interval);
+  }, [isButtonEnabled]);
+
+  const handleCancel = async () => {
+    try {
+      // Make sure the min life isnt displayed when button is clicked
+      setInProgress(true);
+      // Make sure the user cannot click again on the button
+      setButtonEnabled(false);
+      await props.stream.cancel();
+      props.onComplete(true);
+    } catch (error) {
+      if (error instanceof Error) {
+        props.onComplete(error.message);
+      } else {
+        props.onComplete(String(error));
+      }
+    }
+  };
 
+  // Fetch starttime
+  useEffect(() => {
+    const getStartTime = async () => {
+      try {
+        const streamInfo = await props.stream.streamInfo();
+        setStartTime(Number(streamInfo.start_time));
+        if (streamInfo) {
+          clearInterval(interval);
+        }
+      } catch (error) {
+        console.error("Error getting stream token");
+      }
+    };
+    const interval = setInterval(getStartTime, 1000);
+    return () => clearInterval(interval);
+  }, [startTime]);
+
+  // Fetch current time every second
+  useEffect(() => {
+    const interval = setInterval(() => {
+      setCurrentTime(Date.now() / 1000);
+    }, 1000);
+
+    return () => clearInterval(interval);
+  }, []);
+
+  // Calculate the time in seconds before a stream can be cancelled
+  const timeBeforeCancellability = startTime + minStreamLife - currentTime;
+
+  return (
+    <div className="stream-container">
+      <div className="cancel-stream-label">
+        {minStreamLife === null ? (
+          <div>Fetching stream minimum life...</div>
+        ) : !isButtonEnabled && !inProgress && startTime !== 0 ? (
+          <div>
+            Deployment cannot be cancelled yet: its minimum life is
```

Comment:
Might want to eventually add an info popup to clue people into why this is the case